### PR TITLE
Add provisional Probe service receipts

### LINKED ISSUE #710: DRanking: versioned service receipts and a provisional probe vertical slice
## Summary

Define a canonical, versioned service-receipt model and land one end-to-end provisional probe-receipt vertical slice.

Before the admission ledger exists, collected receipts are provisional evidence observed by local nodes. They are not finalized DRanking input, do not create rank, and must not change order_peers_by_quality. When ledger-delimited epochs land, the network performs a deliberate receipt-protocol cutover; provisional time-slice receipts are not silently reinterpreted as ledger receipts.

The first vertical slice is probe because its request, authenticated response, unit count, and beneficiary acknowledgement can be defined without claiming durable storage or hidden-path relay delivery. Relay and store require separate evidence profiles after the common carrier is proven.

## Security claim

A valid provisional receipt proves that:

- the provider and beneficiary account identities are fixed by delegated session proofs;
- both delegated session keys signed the same canonical claim in distinct role domains;
- the collector accepted the acknowledgement while both attestations and the provisional epoch were live;
- the receipt has not already been admitted under its epoch-scoped nonce key.

It does not prove:

- external truth beyond the signed probe transcript;
- that a provider stayed live after the response;
- ledger inclusion or finality;
- globally unique issuance from the receipt bytes alone;
- Sybil resistance, honest service value, or trusted weight;
- that a session-signed receipt observed before the ledger remains safe against arbitrary historical backfilling after delegated-key compromise.

The last limitation is why these receipts remain provisional. Future ledger admission supplies the ordering/finality boundary that this issue does not have.

## Canonical model

Add explicit v1 provisional types in rings-core:

    ProvisionalServiceClaimV1 {
        network_id: u32,
        service: ServiceKindV1,
        provider_account: Did,
        beneficiary_account: Did,
        epoch: ProvisionalEpochV1,
        nonce: [u8; 32],
        units: NonZeroU64,
        request_digest: [u8; 32],
        completion_digest: [u8; 32],
    }

    ProvisionalServiceReceiptV1 {
        claim: ProvisionalServiceClaimV1,
        provider_attestation: MessageVerification,
        beneficiary_attestation: MessageVerification,
    }

For this issue, ServiceKindV1 admits only Probe. Do not put unimplemented Relay or Store variants on the accepted wire and then assign them placeholder verification semantics.

The canonical claim digest binds every field. Provider and beneficiary signatures use distinct, explicit domains so the signatures cannot be exchanged:

    rings-core:service-receipt:provisional:provider:v1
    rings-core:service-receipt:provisional:beneficiary:v1

The receiver supplies network_id to verification, following the existing message-domain discipline. The claim also carries network_id so its portable canonical bytes are self-describing; a receipt from another overlay must fail both structural and signature-domain validation.

units is mandatory. For ProbeV1 it is exactly 1. Zero, any other value, arithmetic overflow, an unknown service kind, or a non-canonical encoding fails closed.

provider_account and beneficiary_account must equal the account DIDs recovered from their respective Session proofs. Session DIDs never identify receipt roles and session rotation does not change the account pair.

## Probe receipt flow

Use a three-step authenticated transcript:

1. Beneficiary b sends a probe request containing a fresh random nonce.
2. Provider a responds with the matching completion plus ProvisionalServiceClaimV1 and provider_attestation.
3. Beneficiary b verifies the response, claim fields, unit rule, request digest, and completion digest, then returns beneficiary_attestation over the same claim.

The provider assembles and stores the receipt only after verifying the beneficiary acknowledgement. A receipt offered by the provider without the beneficiary acknowledgement is incomplete and is never admitted as evidence.

The request and completion digests are canonical signed-transaction digests, not hashes of ad hoc debug or JSON representations. They bind the receipt to the exact request/response exchange. Role checks must reject self-inconsistent transcripts even when every individual signature is otherwise valid.

Replace the current timestamp-only PeerLivenessProbe/PeerLivenessReport wire shape as one breaking cutover. Do not keep a dual decoder or optional legacy fields. Timing may remain local reliability evidence, but a sender-provided timestamp is not receipt freshness.

## Provisional epochs

Define one published constant width:

    PROVISIONAL_RECEIPT_EPOCH_SECS

and:

    ProvisionalEpochV1 {
        slot: u64,
    }

slot is the coarse Unix-time slice agreed at live collection. It is not a ledger height. Provider and beneficiary reject a claim outside the locally accepted current/adjacent-slot tolerance before signing; the collector checks the same rule before storing.

The provisional signature domains make these claims cryptographically distinct from future ledger receipts. When #712 supplies finalized epochs, ledger receipts use a new LedgerEpochId and new signature domains. There is no dual interpretation, compatibility bridge, or automatic import of provisional receipts into P_k. Any later genesis/import attestation is a separate explicit protocol decision.

## Three verification layers

Keep these predicates separate.

### Cryptographic verification

    verify_crypto(receipt, receiver_network_id)

This can be run later by a third party holding the receipt bytes. It checks canonical decoding, supported version/kind, network binding, both claim signatures, role/account equality, unit rules, and request/completion digest shape.

It does not claim nonce freshness or timely observation.

### Live collection admission

    admit_observed(state, receipt, observed_at)

This checks verify_crypto, current MessageVerification liveness, provisional-epoch tolerance, nonce freshness, and storage bounds before retaining the receipt. The freshness key is:

    (network_id, beneficiary_account, epoch, nonce)

A duplicate key is rejected even if the remaining claim fields differ. That conflict is observable evidence, not a second receipt.

### Future ledger aggregation

Freshness inside a finalized P_k and deterministic w_max clipping are set-relative operations owned by #711/#712. #710 must define the canonical receipt digest and ordering required by those operations, but must not pretend one receipt can prove its own uniqueness or issuance cap.

The phrase cross-epoch nonce state is zero applies to aggregation: P_k deduplicates its own epoch set. It does not mean a live collector may discard all bounded replay/storage state without changing its local admission guarantee.

## rings-measure integration

Store verified provisional receipts in a separate bounded evidence component. Do not insert variable receipt collections into CreditRecord and do not rewrite existing byte counters as if they were receipt-derived.

The current local credit and reliability values remain direct local observations. order_peers_by_quality remains unchanged until signer-trust weighting and finalized aggregation exist; otherwise arbitrary low-cost identities could countersign each other into routing preference before DRanking supplies the defense.

Required evidence-store properties:

- hard global and per-account-pair record/byte bounds;
- deterministic duplicate/conflict handling by canonical receipt digest and freshness key;
- bounded page/query APIs rather than unbounded snapshots;
- native and IndexedDB persistence with the same canonical representation;
- explicit eviction reporting and counters;
- malformed or unverifiable records cannot hide valid records during projection.

Eviction affects evidence availability, not local CreditRecord totals and not any current routing decision.

## Layering

- rings-core: canonical receipt/claim types, domains, digests, pure verification predicates, and ProbeV1 messages.
- rings-measure: bounded provisional evidence storage and projection; no trust scoring.
- node/browser providers: storage and message-flow wiring only.
- #711: canonical epoch-set aggregation, deduplication, units-to-value conversion, w_max clipping, and trust algebra input.
- #712: finalized epoch identity, receipt admission into P_k, ordering, and finality.

## Protocol cutover

This is a breaking protocol change:

- replace the old liveness probe/report wire shape;
- accept only the new ProbeV1 transcript;
- do not provide a dual decoder, feature flag, downgrade path, or legacy fallback;
- update native, relay, browser/Wasm, examples, deterministic simulations, and documentation in the same release.

Mixed-version overlays are unsupported.

## Acceptance criteria

- [ ] Add the canonical provisional claim and receipt types with explicit network, roles, epoch, nonce, units, and request/completion digests.
- [ ] Add distinct provider and beneficiary signature domains and reject role-swapped signatures.
- [ ] Implement the three-step ProbeV1 request, provider offer, and beneficiary acknowledgement flow.
- [ ] Require units == 1 for ProbeV1 and reject zero, unsupported kinds, and overflow.
- [ ] Separate byte-only cryptographic verification from live collection admission and future set-relative aggregation.
- [ ] Reject tampering with network, account roles, epoch, nonce, units, or either transcript digest.
- [ ] Reject a repeated (network, beneficiary, epoch, nonce) key without admitting a second receipt.
- [ ] Prove session rotation preserves account roles while a mismatched delegated account fails.
- [ ] Add deterministic canonical-encoding and digest golden vectors on native and Wasm.
- [ ] Add bounded native and IndexedDB restart tests for a non-empty evidence store.
- [ ] Keep CreditRecord, local reliability, and order_peers_by_quality behavior unchanged.
- [ ] Expose bounded evidence queries plus aggregate counters without account DIDs as metrics labels.
- [ ] Remove legacy probe decoding and update all protocol consumers in one cutover.
- [ ] Document the provisional/finalized boundary and non-claims in SECURITY.md and the DRanking documentation.
- [ ] Open separate evidence-profile issues before enabling Relay or Store receipts.

## Non-goals

- Relay or Store receipt semantics in this first vertical slice.
- Ledger-finalized receipt epochs or P_k admission.
- Importing provisional receipts into the future ledger.
- Applying w_max or trust weighting to current peer ordering.
- Account-key signing on every service event; the runtime currently owns delegated session keys.
- Proof of physical-world service, future liveness, or storage availability.
- Payments, transferable rewards, slashing, or reputation policy.

Ref: papers/dranking.tex, Service Receipts and Input Block; #711 (aggregation/trust algebra); #712 (ledger/finality).

### PR DESCRIPTION
**Pull request by zkjoie**
### Please check if the PR fulfills these requirements

- [x] Tests for the changes have been added
- [x] Docs have been added or updated

## What kind of change does this PR introduce?

Protocol feature, bounded durable evidence persistence, executable formal models, and documentation.

## What is the current behavior?

Probe liveness uses timestamp-only `PeerLiveness` messages and does not produce portable, jointly signed service evidence. See #710.

## What is the new behavior?

- Defines canonical `ProvisionalServiceClaim` and `ProvisionalServiceReceipt` types with distinct provider and beneficiary signing domains.
- Replaces the legacy liveness wire with the three-step `ProbeRequest`, `ProbeOffer`, and `ProbeAcknowledgement` flow.
- Separates pure receipt transition planning from clock, sequence, pending-state, send, and persistence effects through an explicit effect interpreter.
- Verifies the embedded transactions and all four delegated-session boundaries at one explicit observation instant.
- Adds bounded receipt storage plus independently bounded durable replay markers. Receipt eviction cannot reopen an active freshness key; marker saturation and clock regression fail closed.
- Commits evidence and its replay marker before admission returns success. Storage failure rolls the transition back, and successful admission survives process restart without an explicit flush.
- Requires browser IndexedDB evidence storage to open successfully; constructors without a durable evidence backend keep liveness but cannot return a successful receipt admission.
- Keeps provisional evidence isolated from `CreditRecord` and route ordering.
- Adds executable Stateright models for time, loss, duplication, reordering, replay, independent session expiry, admission outcomes, pair/global/per-beneficiary bounds, eviction, clock regression, restart normalization, and hard-crash recovery. Production-refinement tests replay the model through real signatures and the real evidence store.

## Breaking change

Yes. This is a deliberate hard cutover: legacy `PeerLiveness` wire variants are removed with no fallback, negotiation, or dual decoder. Mixed-version peers are unsupported. The workspace and npm package versions are bumped to `0.25.0`.

## Validation

- `rings-measure`: 48 passed
- `rings-core` all features: 811 passed, 4 ignored; doc tests 1 passed, 2 ignored
- `rings-node` all features: 358 passed, 1 network test ignored; CLI target 7 passed
- Native evidence restart succeeds without an explicit flush; missing durable evidence storage fails closed
- Strict Clippy passes for `rings-measure`, `rings-core`, native `rings-node`, core Wasm, browser lib, and browser tests
- `rings-core` rustdoc passes with warnings denied
- Formatting, patch checks, typos, and locked root/frontend metadata pass

Closes #710

### CODE REVIEW

**Comment by r2d2-rs** [comment]
Required follow-up evidence profiles are now tracked separately: #765 (Relay) and #766 (Store). This PR continues to accept only ProbeV1.

File: crates/core/src/dht/stabilization.rs [hunk 1]
```diff
@@ -585,14 +586,31 @@ impl Stabilizer {
...
+        let unix_seconds = u64::try_from(now_ms).unwrap_or(0) / 1_000;
         let candidates = self.transport.liveness_probe_candidates(now_ms)?;
         for attempt in candidates {
             let peer = attempt.peer();
             let state = self
                 .transport
                 .get_connection(peer)
                 .map(|conn| conn.webrtc_connection_state());
-            let msg = Message::PeerLivenessProbe(PeerLivenessProbe { sent_at_ms: now_ms });
+            let mut nonce = [0_u8; 32];
```
Comment (RyanKung) [inline crates/core/src/dht/stabilization.rs]: use help function instead

File: crates/core/src/dht/stabilization.rs [hunk 2]
```diff
@@ -585,14 +586,31 @@ impl Stabilizer {
...
+            let payload = self
+                .transport
+                .signed_payload(Message::ProbeRequestV1(request), peer, peer)
+                .await?;
+            let tx_id = payload.transaction.tx_id;
+            if !self
+                .transport
+                .register_pending_liveness_probe(attempt, tx_id, request)?
+            {
+                continue;
```
Comment (RyanKung) [inline crates/core/src/dht/stabilization.rs]: use filter, not `continue`
code in more functional way

File: crates/core/src/dht/stabilization.rs [hunk 3]
```diff
@@ -601,19 +619,23 @@ impl Stabilizer {
                 idle_ms = PEER_LIVENESS_IDLE_MS,
                 "STABILIZATION peer liveness probe send start"
             );
-            match self.transport.send_direct_message(msg, peer).await {
-                Ok(tx_id) => {
-                    self.transport
-                        .record_peer_liveness_probe_sent(attempt, now_ms)?;
+            match self.transport.send_payload(payload).await {
+                Ok(()) => {
+                    let pending = self
```
Comment (RyanKung) [inline crates/core/src/dht/stabilization.rs]: bad var name. what is pending means? is pending? pending list?

File: crates/core/src/message/handlers/connection.rs [hunk 4]
```diff
@@ -30,34 +28,6 @@ use topology_view::connect_successor_hint;
 #[cfg(all(test, not(target_family = "wasm")))]
 use topology_view::topology_has_confirmed_peer;
 
-/// PeerLivenessProbe is a direct overlay liveness probe.
-#[cfg_attr(all(feature = "wasm", target_family = "wasm"), async_trait(?Send))]
-#[cfg_attr(not(all(feature = "wasm", target_family = "wasm")), async_trait)]
-impl HandleMsg<PeerLivenessProbe> for MessageHandler {
```
Comment (RyanKung) [inline crates/core/src/message/handlers/connection.rs:36]: This is a break change, version bump is required

File: crates/core/src/message/handlers/service_receipt.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+use crate::message::ProvisionalServiceClaimV1;
+use crate::message::ProvisionalServiceReceiptV1;
+use crate::message::ServiceReceiptError;
+use crate::message::Transaction;
+use crate::utils::get_epoch_ms;
+
+/// ProbeV1 request is a direct authenticated overlay liveness probe.
+#[cfg_attr(all(feature = "wasm", target_family = "wasm"), async_trait(?Send))]
+#[cfg_attr(not(all(feature = "wasm", target_family = "wasm")), async_trait)]
+impl HandleMsg<ProbeRequestV1> for MessageHandler {
```
Comment (RyanKung) [inline crates/core/src/message/handlers/service_receipt.rs]: is `V1` label necessory? if we have v2, then we can flag it as v2, but why we now have `v1`

File: crates/core/src/message/handlers/service_receipt.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    async fn handle(&self, ctx: &MessagePayload, msg: &ProbeRequestV1) -> Result<()> {
+        if ctx.should_forward_from(self.dht.did) {
+            return self
+                .run_effects([CoreEffect::forward_payload(ctx, None)])
+                .await;
+        }
+
+        let now_seconds = u64::try_from(get_epoch_ms() / 1_000)
+            .map_err(|_| ServiceReceiptError::ObservationTimeOverflow)?;
+        if !msg.epoch.is_accepted_at(now_seconds) {
```
Comment (RyanKung) [inline crates/core/src/message/handlers/service_receipt.rs]: use match { foo => cond_a(), bar=> cond_b() } ,do not do early return

File: crates/core/src/message/handlers/service_receipt.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            .into());
+        }
+        let beneficiary = ctx.transaction.origin();
+        let provider = self.dht.did;
+        let network_id = self.transport.network_id;
+        let request_digest = ctx.transaction.digest()?.into_bytes();
+        let completion_sequence = *self
+            .transport
+            .reserve_transaction_sequences(beneficiary, NonZeroU64::MIN)
+            .await?
```
Comment (RyanKung) [inline crates/core/src/message/handlers/service_receipt.rs]: is this a side-effect? do not mix side-effect and pure func.
abstract sideEffect with monad

File: crates/core/src/message/handlers/service_receipt.rs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            ctx,
+            self.transport.network_id,
+            beneficiary,
+            get_epoch_ms(),
+        )?;
+        if !self
+            .transport
+            .consume_pending_probe(provider, ctx.transaction.tx_id, request)?
+        {
+            return Err(Error::InvalidMessage(
```
Comment (RyanKung) [inline crates/core/src/message/handlers/service_receipt.rs]: you should map it to error, but dont do early return

File: crates/core/src/message/service_receipt/model.rs [hunk 9]
```diff
@@ -0,0 +1,503 @@
+//! Executable finite-state model for the provisional ProbeV1 receipt flow.
```
Comment (RyanKung) [inline crates/core/src/message/service_receipt/model.rs]: what is this mod for， if it's a formal model it should be ignored when we compile product code

**Review by RyanKung — CHANGES_REQUESTED** [review]

File: crates/core/src/dht/stabilization.rs (hunk 1 above)
Comment (r2d2-rs) [reply crates/core/src/dht/stabilization.rs]: Fixed in `126156f4`: probe nonce and epoch construction now goes through `ProbeRequest::random_for_epoch` instead of open-coded random-byte setup in stabilization.

File: crates/core/src/dht/stabilization.rs (hunk 2 above)
Comment (r2d2-rs) [reply crates/core/src/dht/stabilization.rs]: Fixed in `126156f4`: candidates now flow through sequential `then` -> `try_filter_map` -> `try_for_each`; superseded generations are filtered without `continue`.

File: crates/core/src/dht/stabilization.rs (hunk 3 above)
Comment (r2d2-rs) [reply crates/core/src/dht/stabilization.rs]: Fixed in `126156f4`: renamed the ambiguous `pending` binding/log field to `matching_probe_recorded`, which states exactly what the boolean means.

File: crates/core/src/message/handlers/connection.rs (hunk 4 above)
Comment (r2d2-rs) [reply crates/core/src/message/handlers/connection.rs:36]: Fixed in `126156f4`: this breaking wire cutover now bumps the Rust workspace, internal path constraints, npm package, root lockfile, and frontend lockfile to `0.25.0`; the changelog records the coordinated-upgrade requirement.

File: crates/core/src/message/handlers/service_receipt.rs (hunk 5 above)
Comment (r2d2-rs) [reply crates/core/src/message/handlers/service_receipt.rs]: Fixed in `126156f4`: removed the `V1` suffix from all public Rust receipt types and `Message` variants. `V1` remains only inside canonical byte markers and signing domains as a cryptographic domain separator.

File: crates/core/src/message/handlers/service_receipt.rs (hunk 6 above)
Comment (r2d2-rs) [reply crates/core/src/message/handlers/service_receipt.rs]: Fixed in `126156f4`: forwarding/local dispatch and epoch admission are expressed as `match` transitions, with no handler early return.

File: crates/core/src/message/handlers/service_receipt.rs (hunk 7 above)
Comment (r2d2-rs) [reply crates/core/src/message/handlers/service_receipt.rs]: Fixed in `126156f4`: pure plan/build/role-validation functions are separated from an explicit `ProbeEffect` algebra and `ProbeEffectInterpreter`. Clock reads, sequence reservation, pending-state mutation, network sends, and durable admission now live only in the interpreter boundary.

File: crates/core/src/message/handlers/service_receipt.rs (hunk 8 above)
Comment (r2d2-rs) [reply crates/core/src/message/handlers/service_receipt.rs]: Fixed in `126156f4`: the pending-probe boolean is converted with `then_some(()).ok_or_else(...)`; the handler no longer uses an early return for this rejection.

File: crates/core/src/message/service_receipt/model.rs (hunk 9 above)
Comment (r2d2-rs) [reply crates/core/src/message/service_receipt/model.rs]: Confirmed and made explicit in `126156f4`: the parent module already used `#[cfg(all(test, not(target_family = "wasm")))]`; the model file now also has the same file-level gate. `stateright` is absent from the normal `rings-core` dependency graph.

**Review by RyanKung — APPROVED** [review]
